contrib/flo-scio_2.12/src/main/scala/com/spotify/flo/contrib/scio/ScioJobSpec.scala

```diff
@@ -0,0 +1,120 @@
+/*-
+ * -\-\-
+ * flo-scio
+ * --
+ * Copyright (C) 2016 - 2018 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.flo.contrib.scio
+
+import com.spotify.flo.TaskBuilder.{F0, F1, F2}
+import com.spotify.flo.contrib.scio.ScioJobSpec.log
+import com.spotify.flo.{FloTesting, TaskId}
+import com.spotify.scio.{ScioContext, ScioResult}
+import org.apache.beam.runners.dataflow.DataflowPipelineJob
+import org.apache.beam.sdk.options.{ApplicationNameOptions, PipelineOptions, PipelineOptionsFactory}
+import org.slf4j.{Logger, LoggerFactory}
+
+class ScioJobSpec[R, S](private val taskId: TaskId,
+                        private val _options: Option[F0[PipelineOptions]] = None,
+                        private val _pipeline: F1[ScioContext, _] = null,
+                        private val _result: F2[ScioContext, ScioResult, R] = null,
+                        private val _success: F1[R, S] = null
+                       ) extends Serializable {
+
+  def options(options: F0[PipelineOptions]): ScioJobSpec[R, S] = {
+    new ScioJobSpec[R, S](taskId, Some(options), _pipeline, _result, _success)
+  }
+
+  def pipeline(pipeline: F1[ScioContext, _]): ScioJobSpec[R, S] = {
+    new ScioJobSpec[R, S](taskId, _options, pipeline, _result, _success)
+  }
+
+  def result[RN](result: F2[ScioContext, ScioResult, RN]): ScioJobSpec[RN, S] = {
+    new ScioJobSpec[RN, S](taskId, _options, _pipeline, result, null)
+  }
+
+  def success[SN](success: F1[R, SN]): SN = {
```

Comment:
The method name here might be a bit confusing. I would call it something like
` def run[SN](onSuccess: F1[R, SN]): SN ` That way it's clear to users that this should be called very last, after setting up the options and pipeline.

This also makes me think that having `run()` and  `success(..)` separate might be even clearer, that would allow adding a `fail(..)` for failure handlers

Reply:
Yeah, the idea was that this job api would declarative and let the operator run the job at a time and place of its choosing. Having a `run()` method makes the api more imperative and implies that the job is immediately executed. Maybe it could be called something else like `done()`. Or maybe it's better to go for the obvious `run()` now and remove it later when/if job execution moves out of the process fn. 🤔

Reply:
I see, if that was your intention from the beginning, then I understand